### DISCUSSION TITLE
Issues with the order of correlated and uncorrelated orbitals in Wannier90+CTQMC calculations

### DISCUSSION THREAD

**Discussion by Gurshidali** [TRIQS/dft_tools#242] [General]
Hi,

I have a few inquiries regarding the DFTTools feature within the TRIQS software, particularly in relation to simulating materials such as SrVO3 and Fe-bcc. I would greatly appreciate your assistance in clarifying these questions:

1. I encountered an issue while running CTQMC on a converted h5 file from a general LDA-based tight-binding Hamiltonian in k-space. Successful CTQMC runs were achieved for the SrVO3 system, but failures occurred when attempting CTQMC for the Fe-bcc system with 9 orbitals (4 uncorrelated orbitals followed by 5 correlated orbitals with t2g and eg splitting obtained from wannier90). The issue may be related to the order of the shells in the input file for the Hkconverter. As an alternative, I tried using the interface to wannier90, but CTQMC only ran for a few iterations before stopping, and it was found that only 4 subspaces (possibly s-orbitals) were identified instead of the expected 1024 subspaces (3d correlated orbitals in Fe-bcc). Currently, I provide the shell description in the order of the orbitals as generated by the wannier90 Hamiltonian. However, CTQMC incorrectly assumes the first shells are correlated orbitals. Is there a way to define the correlated shells in the seedname.inp file according to our specific requirements?

2. Is it necessary to use the same high Monte Carlo parameter values (n_warmup_cycles, n_cycles, length_cycle) for sequential finite temperature calculations (different beta values)? Are there any conditions or relationships between beta and the MC parameters that allow for a more systematic increase in MC values to obtain noiseless data and faster convergence?

3. In the context of simulations for real materials, what would be the minimum threshold value to consider for the average sign?

P.S.: Wannier90 always gives the Hamiltonian matrix in the following order: s,p,d (in the case of Fe-bcc) irrespective of however I have mentioned the projections, while wannier90_converter treats the first shell as correlated and the subsequent shells as uncorrelated and which is causing the problem. I'm uncertain about the INP file, particularly in the General H(k) section, where the third line requests the total number of atomic shells, whereas the seedname.inp file asks for the total number of atoms.

Thank you very much for your time and expertise. I look forward to your response.

![Fe-bcc_wannier90_converter](https://github.com/TRIQS/dft_tools/assets/103921975/e52745a0-51c7-4a3c-a9ce-2ee51dfddca8)
![seedname inp](https://github.com/TRIQS/dft_tools/assets/103921975/e0100a60-726f-415a-beb9-e9d2158d4cee)
![General_Hk](https://github.com/TRIQS/dft_tools/assets/103921975/1b054f8c-0e89-4c9e-8c07-91c4fcc19b34)

![DMFT_error](https://github.com/TRIQS/dft_tools/assets/103921975/75a05ed8-acd5-437a-a512-9484658bda66)

With regards,
Gurshid

**Comment by the-hampel** [comment]
Hi @Gurshidali, 
1. Can you share the .inp file you used? The converter output says it only found one correlated WF and treated the 8 remaining WFs as uncorrelated. Note that the converter can only treat the first x WFs a correlated. All uncorrelated WFs need to be last in the wannier90 hr.dat file. It also seems that there are large complex elements in your wannier Hamiltonian. Can you check if your wannier90_hr.dat reproduces your band structure from DFT? This explains why CTQMC only found 4 subspaces, because the converter found only one correlated orbital. Your .inp file might look like this to treat 5 d orbitals as correateld and the remaining 4 orbitals as uncorrelated (need to be last in w90)
```
0 7 7 7
4.0 # specify filling
1
0 0 2 5 0 0
```

2. In general you have to adapt the number of MC cycles, warmup, cycle length, n_tau as you change temperature. We created a new tutorial for this on the unstable branch of cthyb: https://triqs.github.io/cthyb/unstable/guide/cthyb_convergence_tests.html
3. The average sign tells how effective the sampling is. A value of 0.5 means that only every second measurement can be considered, i.e. you need to sample twice as much. Usually I would say that everything between 0.5 - 1.0 is okayish. But if possible think about possible change of basis to improve the sign, e.g. diagonalizing the local impurity Hamiltonian

Let me know if that helps or you have further questions. 

Best,
Alex

**Reply by Gurshidali** [reply]
Hi Alex,

Thank you so much for the informative explanation. You can find the attached .inp file below. I have grasped the answers to the last two questions. However, I still seek further clarity on my initial query. From what I understand, the wannier90 converter is currently restricted to considering only the first set of orbitals as correlated orbitals. The issue was actually in the content of the wannier90_hr.dat file, where uncorrelated orbitals occupy the first positions, followed by the formation of the next five correlated orbitals. 

I am pleased to inform you that I have successfully generated the wannier90_hr.dat file with the desired orbital ordering and also converted it into a .h5 file. However, I must note that the issue with large complex components still persists (hence CTQMC failed to run), despite achieving a good match between wannier90 and DFT bandstructure. For your convenience, I have attached a comparison for your reference.

I kindly request your assistance in this matter. Your valuable suggestions and insights on how to eliminate the large complex component values in the wannier90_hr.dat file would be highly appreciated. 

0 7 7 7
4.0 # specify filling
1
0 0 2 5 0 0

Additionally, In the above-given .inp file: The 3rd line, is it the number of atoms or atomic shells? I am inquiring about this because when I use the value 3 instead of 1 in my .inp file, the converter considers all nine orbitals as correlated orbitals.

Thank you
With regards,
Gurshid

![Screenshot from 2023-07-19 15-45-00](https://github.com/TRIQS/dft_tools/assets/103921975/1b2fdda3-3593-4dc3-ae41-77b27e6531ba)

![DFT-W90-BS](https://github.com/TRIQS/dft_tools/assets/103921975/b53e7bcf-ffcf-497a-b281-ac410b5f5128)

![Screenshot from 2023-07-19 15-25-32](https://github.com/TRIQS/dft_tools/assets/103921975/aee3e25a-054f-433b-bd58-949aff3a4582)

![Screenshot from 2023-07-19 15-53-34](https://github.com/TRIQS/dft_tools/assets/103921975/c4c29250-b75d-4b42-b882-18fe61ef8ab6)

**Reply by the-hampel** [reply]
I think your wannier bands show that there might be still a problem with the wannierization. Especially the lowest and highest band look a bit "wonky" to me. Also I am a bit confused about the number of states. It almost seems as if you have more wannier bands as DFT bands? Are the bands you are trying to wannierize entangled with other KS states? If so how did you choose the disentanglement windows? I believe you can get a better fit to the DFT bands and higher quality WFs. 

> The 3rd line, is it the number of atoms or atomic shells?

it is the number of atomic sites / atoms. It has to match the number of the following lines.

**Reply by Gurshidali** [reply]
Hi Alex,

Thank you for your valuable suggestion about the WFs. The DFT bandstructure I shared was adjusted by eliminating higher and lower bands to compare it with the 9-band W90 bandstructure needed for our Hamiltonian. The KS states indeed show high entanglement, as seen in the plots. I'm currently working on fine-tuning the disentanglement windows based on your commands.

Continuing with the sign problem, I'm concerned that changing the basis to the eigenbasis of the local impurity Hamiltonian might require rotating the interaction term. This could complicate the problem, converting the density-density type to a more general interaction form, like the tensor form, especially since I'm working with a density-density type Hamiltonian. If the interactions are complicated, are we guaranteed to have much less sign problem? And are there any features to diagonalize the local impurity Hamiltonian in TRIQS/DFTTools?

Please let me know your thoughts on this and any additional insights or suggestions from you would be greatly appreciated.

With regards,
Gurshid

**Reply by the-hampel** [reply]
Yes you are absolutely right. You would have to be careful about rotating the interaction Hamiltonian into the same basis. There are functions within triqs that can take care of that. You can find an example of a typical workflow in solid_dmft in the file `interaction_hamiltonian.py`: 
https://github.com/TRIQS/solid_dmft/blob/247183077eed3ee8b4e8cbcda734b3a322a6f43c/python/solid_dmft/dmft_tools/interaction_hamiltonian.py#L452
This is the line that rotates the interaction Hamiltonian with an arbitrary rotation matrix: 
https://github.com/TRIQS/solid_dmft/blob/247183077eed3ee8b4e8cbcda734b3a322a6f43c/python/solid_dmft/dmft_tools/interaction_hamiltonian.py#L396
```
from triqs.operators import util
Umat_full_rotated[icrsh] = util.transform_U_matrix(Umat_full[icrsh], sum_k.rot_mat[ish].T)
```

If you run the wannier90 converter the rotation matrix from the original wannier basis to the diagonal one is stored in the h5 archive as `rot_mat` and can be accessed later in dft_tools in the sumk object via `sumk.rot_mat` for each impurity site. 

You can also see that for certain interaction Hamiltonian's the symmetry is such that you do not need to transform for example with the Kanamori Hamiltonian in its rotational symmetric form. 

Typically we find that using the basis that diagonalizes the non-interacting local Hamiltonian gives the best sign. Even though the interaction Hamiltonian becomes more complicated. But you should carefully test this for each system at hand.

Best,
Alex

**Reply by Gurshidali** [reply]
Dear Alex,

Thank you very much for your prompt response and for providing the necessary clarifications. I will certainly review the information you shared.

regards,
Gurshid
